### PR TITLE
Feature/unit conversion revisited

### PR DESCRIPTION
**Pull request by jfcorbett**
Unit conversions, reimplemented. 

Three clearly separated concerns. Each of them is user-defined... though a default is supplied for the first one. 
	
1. A general unit converter callable: takes a value, a from_unit, a to_unit... spits out the converted value. A default pint-based callable-class converter is supplied (`pint` install is only required upon first usage). User can subclass it if they want to supplement it with their own unit definitions (this is demoed in `test_units.py`). User can also implement their own from scratch by hand, or wrap `unum` or some different package or whatever. 

2. For one given table: a `ColumnUnitDispatcher` that tells the table which columns to convert to what unit. Can be a callable, dict, or list. 

3. For bulk unit conversion of all tables at read time (originally @thpetsen 's idea, this is not a use case we have seen, but it could actually be quite useful): A `TableUnitDispatcher` which is a callable or dict that returns `ColumnUnitDispatcher`s based on table name. 

This is many furlongs per fortnight better than what was in the old `startables` package. 

Still missing:
- dealing with NaNs and missing values such in table columns
- convenience methods to "convert back to origin units" without having to explicitly specify them anew
- ... what else?

### CODE REVIEW

**Comment by jfcorbett** [comment]
Oops, flake8 doesn't like "pint.Unit" in type hints because 'pint' isn't imported at compile time. I'll move the DefaultUnitConverter around a bit to fix this.

File: environment.yml [hunk 1]
```diff
@@ -10,11 +10,13 @@ dependencies:
   - pandas
 # OPTIONAL
```
Comment (thpetsen) [inline environment.yml:11]: Perhaps # comment these out

File: pdtable/demo/unit_converter.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            Old unit
+        to_unit:
+            New unit to which to convert
+
+    Returns:
+        Number or array converted from old to new unit.
+
+    """
+    requested_conversion = (from_unit, to_unit)
+    available_conversions = {
```
Comment (thpetsen) [inline pdtable/demo/unit_converter.py:27]: We have much more than this if we separate magnitude and type (nano, micro, centi, hekto, kilo, mega &c.)

File: pdtable/proxy.py [hunk 3]
```diff
@@ -237,17 +260,53 @@ def equals(self, other):
...
+                    col.convert_units(to_unit, converter)
+
+        elif isinstance(to, Dict):
+            for column in self.column_proxies:
+                to_unit = to.get(column.name)
+                if to_unit is not None:
+                    column.convert_units(to[column.name], converter)
+
+        else:
+            # Assume 'to' is callable
```
Comment (thpetsen) [inline pdtable/proxy.py:305]: better assert => easier to understand error message

File: pdtable/test/test_units.py [hunk 4]
```diff
@@ -1,106 +1,133 @@
...
 
 
-@pytest.fixture
-def unit_policy() -> UnitPolicy:
-    return SuperSimpleUnitPolicy()
+class CustomUnitConverter(DefaultUnitConverter):
+    def __init__(self):
+        super().__init__()
 
+    def __call__(self, value, from_unit, to_unit):
```
Comment (thpetsen) [inline pdtable/test/test_units.py:35]: Good idea ! Allows configuring converters &c. should be included as an example.

**Review by thpetsen — APPROVED** [review]

File: pdtable/test/test_units.py (hunk 4 above)
Comment (jfcorbett) [reply pdtable/test/test_units.py:35]: Now https://github.com/startable/pdtable/issues/50

File: environment.yml (hunk 1 above)
Comment (jfcorbett) [reply environment.yml:11]: They are also needed in test environment. 

Incidentally, they are only "optional" for the end user, until first usage of Excel reader/writer and of default unit converter, respectively.

Note: this environment.yml file defines the development environment. It won't be used at the time of packaging, and it won't determine the dependencies of the pip or conda packages.

File: pdtable/demo/unit_converter.py (hunk 2 above)
Comment (jfcorbett) [reply pdtable/demo/unit_converter.py:27]: that's true, but this is just a simple demo sketch. Not trying to be complete, just illustrating the idea. 

Now that we have the DefaultUnitConverter, I even thought of removing it entirely... but it's still useful to show that you can write your own by hand if you want.

File: pdtable/proxy.py [hunk 5]
```diff
@@ -304,13 +304,15 @@ def convert_units(self, to: ColumnUnitDispatcher, converter: UnitConverter):
                 if to_unit is not None:
                     column.convert_units(to[column.name], converter)
```
Comment (JanusWesenberg) [inline pdtable/proxy.py:306]: I would vote for skipping the instance check in favor of decorating and forwarding any exceptions raised? Same below for utils.py

**Review by JanusWesenberg — APPROVED** [review]
Just a thought: We could decouple the table class from the whole unit handling if we took a functional approach instead of updating units in place.

File: pdtable/proxy.py (hunk 5 above)
Comment (jfcorbett) [reply pdtable/proxy.py:306]: @JanusWesenberg  I think I need an example... what do you mean by "decorating"? 

I don't like `isinstance()` checks and I'm sure there's a better alternative, but I don't know what it is.

**Comment by jfcorbett** [comment]
@JanusWesenberg What would a functional API look like? 
```
t2 = convert_table_units(table, to={"diameter": "cm"}, converter=whatever)
```
Something like this? What do we gain with the former? Instead of the existing
```
table.convert_units(to={"diameter": "cm"}, converter=whatever)
# could even return self to allow:
t2 = table.convert_units(to={"diameter": "cm"}, converter=whatever)
```
... which I find more readable.

**Comment by JanusWesenberg** [comment]
> @JanusWesenberg What would a functional API look like?
> 
> ```
> t2 = convert_table_units(table, to={"diameter": "cm"}, converter=whatever)
> ```

Maybe more along the lines of 
```
my_normalizer = BaseUnitNormalize(...)
table_normalized = my_normalizer(table)
```

One of the nice aspects is that tables would not need to know about this at all: they are just constructed. This also means that we can supply some normalizer tools that make sense for the use cases we know of, but then other projects can build their own.

> Something like this? What do we gain with the former? Instead of the existing
> ...
> ... which I find more readable.

I could easily imagine a situation where table objects are shared, in which case in-place modifications of values can lead to all sorts of subtle bugs. The take in pandas seems to be that no in-place happens unless there is an explicit `in-place` argument.

**Comment by jfcorbett** [comment]
Good point with shared tables... 
I'm running out of steam for this PR, so let's just complete it and return to this:
Created Issue https://github.com/startable/pdtable/issues/54

**Pull request merged**